daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/LimitsSpec.scala

```diff
@@ -0,0 +1,436 @@
+          signatories Mod:T {signatories} this;
+          observers Mod:T {observers} this;
+          agreement "Agreement";
+          choice @nonConsuming NoOp (self) (arg: Mod:NoOpArg): Unit,
+            controllers Mod:NoOpArg {controllers} arg,
+            observers Mod:NoOpArg {observers} arg
+            to
+              upure @Unit ();
+         };
+      }
+    """
+  )
+
+  "Machine" should {
+
+    val committers = (0 to 99).view.map(i => Ref.Party.assertFromString(s"Parties$i")).toSet
+
+    "refuse to build contract with too many signatories" in {
+      val limit = 10
+      val limits = interpretation.Limits.Lenient.copy(contractSignatories = limit)
+      val e =
+        e"""\(signatories: List Party) (observers: List Party) -> 
+             create @Mod:T Mod:T { signatories = signatories, observers = observers }
+         """
+
+      List(1, limit, limit + 1, 99).count { i =>
+        val (signatories, observers) = committers.splitAt(i)
+        val result =
+          eval(limits, Map.empty, signatories, e, asSParties(signatories), asSParties(observers))
+        if (i <= limit) {
+          result shouldBe a[Right[_, _]]
+          true
+        } else {
+          inside(result) {
+            case Left(
+                  SError.SErrorDamlException(
+                    IE.Limit(IE.Limit.ContractSignatories(_, templateId, _, limit))
+                  )
+                ) =>
+              templateId shouldBe T
+              limit shouldBe limit
+              false
+          }
+        }
+      } shouldBe 2
+    }
+
+    "refuse to fetch contract with too many signatories" in {
+      val limit = 10
+      val limits = interpretation.Limits.Lenient.copy(contractSignatories = limit)
+      val e = e"""\(cid: ContractId Mod:T) -> fetch @Mod:T cid"""
+
+      List(1, limit, limit + 1, 49).count { i =>
+        val (signatories, observers) = committers.splitAt(i)
+        val contract = mkContract(signatories, observers)
+        val result = eval(limits, Map(aCid -> contract), signatories, e, SValue.SContractId(aCid))
+        if (i <= limit) {
+          result shouldBe a[Right[_, _]]
+          true
+        } else {
+          inside(result) {
+            case Left(
+                  SError.SErrorDamlException(
+                    IE.Limit(IE.Limit.ContractSignatories(_, templateId, _, limit))
+                  )
+                ) =>
+              templateId shouldBe T
+              limit shouldBe limit
+              false
+          }
+        }
+      } shouldBe 2
+    }
+
+    "refuse to exercise contract with too many signatories" in {
+      val limit = 10
+      val limits = interpretation.Limits.Lenient.copy(contractSignatories = limit)
+      val e =
+        e"""\(cid: ContractId Mod:T) (controllers: List Party) -> 
+           exercise @Mod:T NoOp cid Mod:NoOpArg {controllers = controllers, observers = Nil @Party }"""
+
+      List(1, limit, limit + 1, 49).count { i =>
```

Comment:
I’m also a bit confused by the numbers here. Is 49 `limit - 1`? If so, should we just write it as that?

Reply:
I replaced by limit * 5.